pkg/authorization/authorizer/scope/converter.go

```diff
@@ -0,0 +1,141 @@
+			}
+		}
+
+		if !found {
+			errors = append(errors, fmt.Errorf("no scope evaluator found for %q", scope))
+		}
+	}
+
+	return rules, kutilerrors.NewAggregate(errors)
+}
+
+var discoveryRules = []authorizationapi.PolicyRule{
+	{
+		Verbs: sets.NewString("get"),
+		NonResourceURLs: sets.NewString(
+			// API discovery/negotiation
+			"/api", "/api/*",
+			"/apis", "/apis/*",
+			"/oapi", "/oapi/*",
+		),
+	},
+}
+
+const (
+	SpecialIndicator       = "special:"
+	ClusterRoleIndicator   = "clusterrole:"
+	ClusterWideIndicator   = "clusterwide:"
+	NamespaceWideIndicator = "namespace:"
+)
+
+// scopeEvaluator takes a scope and returns the rules that express it
+type scopeEvaluator func(scope, namespace string, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) ([]authorizationapi.PolicyRule, error)
+
+// scopeEvaluators map prefixes to a function that handles that prefix
+var scopeEvaluators = map[string]scopeEvaluator{
+	SpecialIndicator:     specialEvaluator,
+	ClusterRoleIndicator: clusterRoleEvaluator,
+}
+
+// scopes are in the format
+// <indicator><indicator choice>
+// we have the following formats:
+// special:<special name>
+// clusterrole:<clusterrole name>:<namespace to allow the cluster role, * means all>
+// TODO
+// cluster:<comma-delimited verbs>:<comma-delimited resources>
+// namespace:<namespace name>:<comma-delimited verbs>:<comma-delimited resources>
+
+const (
+	SpecialWhoAmI      = "whoami"
+	SpecialAccessCheck = "access-check"
+)
+
+// special:<special name>
+func specialEvaluator(scope, namespace string, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) ([]authorizationapi.PolicyRule, error) {
+	switch scope {
+	case SpecialIndicator + SpecialWhoAmI:
+		return []authorizationapi.PolicyRule{
+			{Verbs: sets.NewString("get"), APIGroups: []string{userapi.GroupName}, Resources: sets.NewString("users"), ResourceNames: sets.NewString("~")},
+		}, nil
+	case SpecialIndicator + SpecialAccessCheck:
+		return []authorizationapi.PolicyRule{
+			{Verbs: sets.NewString("create"), Resources: sets.NewString("subjectaccessreviews", "localsubjectaccessreviews"), AttributeRestrictions: &authorizationapi.IsPersonalSubjectAccessReview{}},
+		}, nil
+	default:
+		return nil, fmt.Errorf("unrecognized scope: %v", scope)
+	}
+}
+
+// clusterrole:<clusterrole name>:<namespace to allow the cluster role, * means all>
+func clusterRoleEvaluator(scope, namespace string, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) ([]authorizationapi.PolicyRule, error) {
+	tokens := strings.SplitN(scope, ":", 2)
+	if len(tokens) != 2 {
+		return nil, fmt.Errorf("bad format for scope %v", scope)
+	}
+
+	// namespaces can't have colons, but roles can.  pick last.
+	lastColonIndex := strings.LastIndex(tokens[1], ":")
+	if lastColonIndex <= 0 || lastColonIndex == (len(tokens[1])-1) {
+		return nil, fmt.Errorf("bad format for scope %v", scope)
+	}
+	roleName := tokens[1][0:lastColonIndex]
+	scopeNamespace := tokens[1][lastColonIndex+1:]
```

Comment:
make sure len(roleName) and len(scopeNamespace) are > 0? Otherwise, empty namespace matches cluster-scoped resource requests